src/spectrum.rs

```diff
@@ -0,0 +1,375 @@
+        (flt(0.011359160000),flt(0.004102000000),flt(0.000000000000)),
+        (flt(0.008110916000),flt(0.002929000000),flt(0.000000000000)),
+        (flt(0.005790346000),flt(0.002091000000),flt(0.000000000000)),
+        (flt(0.004109457000),flt(0.001484000000),flt(0.000000000000)),
+        (flt(0.002899327000),flt(0.001047000000),flt(0.000000000000)),
+        (flt(0.002049190000),flt(0.000740000000),flt(0.000000000000)),
+        (flt(0.001439971000),flt(0.000520000000),flt(0.000000000000)),
+        (flt(0.000999949300),flt(0.000361100000),flt(0.000000000000)),
+        (flt(0.000690078600),flt(0.000249200000),flt(0.000000000000)),
+        (flt(0.000476021300),flt(0.000171900000),flt(0.000000000000)),
+        (flt(0.000332301100),flt(0.000120000000),flt(0.000000000000)),
+        (flt(0.000234826100),flt(0.000084800000),flt(0.000000000000)),
+        (flt(0.000166150500),flt(0.000060000000),flt(0.000000000000)),
+        (flt(0.000117413000),flt(0.000042400000),flt(0.000000000000)),
+        (flt(0.000083075270),flt(0.000030000000),flt(0.000000000000)),
+        (flt(0.000058706520),flt(0.000021200000),flt(0.000000000000)),
+        (flt(0.000041509940),flt(0.000014990000),flt(0.000000000000)),
+        (flt(0.000029353260),flt(0.000010600000),flt(0.000000000000)),
+        (flt(0.000020673830),flt(0.000007465700),flt(0.000000000000)),
+        (flt(0.000014559770),flt(0.000005257800),flt(0.000000000000)),
+        (flt(0.000010253980),flt(0.000003702900),flt(0.000000000000)),
+        (flt(0.000007221456),flt(0.000002607800),flt(0.000000000000)),
+        (flt(0.000005085868),flt(0.000001836600),flt(0.000000000000)),
+        (flt(0.000003581652),flt(0.000001293400),flt(0.000000000000)),
+        (flt(0.000002522525),flt(0.000000910930),flt(0.000000000000)),
+        (flt(0.000001776509),flt(0.000000641530),flt(0.000000000000)),
+        (flt(0.000001251141),flt(0.000000451810),flt(0.000000000000)),
+    ]
+}
+
+#[inline(always)]
+fn lerp<T: Float>(a: T, b: T, t: T) -> T {
+    a + (b - a) * t
+}
+
+/// Spectrum representing the relative intensity of 
+/// wavelengths from 360 nm to 830 nm.
+pub struct Spectrum<T: Float> {
+    data: [T; SPECTRUM_SAMPLES],
+}
+
+impl<T: Float> Clone for Spectrum<T> {
+    fn clone(&self) -> Spectrum<T> {
+        let mut data: [T; SPECTRUM_SAMPLES] = [T::zero(); SPECTRUM_SAMPLES];
+        for (sample, cloned) in self.data.iter().zip(data.iter_mut()) {
+            *cloned = *sample;
+        }
+        Spectrum {
+            data: data,
+        }
+    }
+}
+
+impl<T> Spectrum<T>
+    where T: Float + Zero + FromPrimitive
+{
+    /// Create a Spectrum from an array of spectral intensity
+    /// values assumed to be in the range 360 nm to 830 nm.
+    pub fn new(data: [T; SPECTRUM_SAMPLES]) -> Result<Self> {
+        if data.iter().any(|&intensity| intensity < T::zero()) {
+            Err(ErrorKind::SpectrumIntensityOutOfRange.into())
```

Comment:
I guess if we were going with what the rest of the library does then we should use `assert!` and expect the caller to make sure there are no invalid values in the array.

I've updated the code to do this and that has removed the need for `error-chain` and the `errors.rs` module which I've also removed.